File: typescript/src/middleware/guardrails/response_validator.ts
import { Nudge, retryNudge, unknownToolNudge } from "./nudge";

export interface ToolCall {
  tool: string;
  args: Record<string, unknown>;
}

export interface ValidationResult {
  toolCalls: ToolCall[];
  nudge: Nudge | null;
  needsRetry: boolean;
}

export class ResponseValidator {
  private toolNames: Set<string>;
  private rescueEnabled: boolean;
  private retryNudgeFn: (rawResponse: string, toolNames: string[]) => Nudge;

  private thinkPattern: RegExp;
  private pythonTagPattern: RegExp;
  private codeFencePattern: RegExp;
  private rehearsalPattern: RegExp;
  private qwenFunctionPattern: RegExp;

  constructor(
    toolNames: string[],
    rescueEnabled: boolean = true,
    retryNudgeFn?: (rawResponse: string, toolNames: string[]) => Nudge,
  ) {
    this.toolNames = new Set(toolNames);
    this.rescueEnabled = rescueEnabled;
    this.retryNudgeFn = retryNudgeFn ?? retryNudge;

    this.thinkPattern = new RegExp("\\[THINK\\].*?\\[/THINK\\]|<think>.*?</think>", "gi");
    this.pythonTagPattern = new RegExp("\\<\\|python_tag\\|\\>", "gi");
    this.codeFencePattern = new RegExp("```(?:json)?\\s*\\n?", "g");
    this.rehearsalPattern = new RegExp("(\\w+)\\[ARGS\\](\\{.*?\\})", "g");
    this.qwenFunctionPattern = new RegExp("<function=([^>\\s]+)>(.*?)<\\/function>", "g");
  }

  validateTextResponse(response: string): ValidationResult {
    if (this.rescueEnabled) {
      const rescued = this.rescueToolCall(response);
      if (rescued.length > 0) {
        return { toolCalls: rescued, nudge: null, needsRetry: false };
      }
    }

    const nudge = this.retryNudgeFn(response, Array.from(this.toolNames));
    return { toolCalls: [], nudge, needsRetry: true };
  }

  validateToolCalls(toolCalls: ToolCall[]): ValidationResult {
    const unknown: string[] = [];
    const validCalls: ToolCall[] = [];

    for (const tc of toolCalls) {
      if (!this.toolNames.has(tc.tool)) {
        unknown.push(tc.tool);
      } else {
        validCalls.push(tc);
      }
    }

    if (unknown.length > 0) {
      const nudge = unknownToolNudge(unknown[0], Array.from(this.toolNames));
      return { toolCalls: [], nudge, needsRetry: true };
    }

    return { toolCalls: validCalls, nudge: null, needsRetry: false };
  }

  private rescueToolCall(response: string): ToolCall[] {
    let cleaned = response.replace(this.thinkPattern, "");
    cleaned = cleaned.replace(this.pythonTagPattern, "");
    cleaned = cleaned.trim();

    if (!cleaned) {
      return [];
    }

    let calls = this.extractJsonToolCalls(cleaned);
    if (calls.length > 0) {
      return calls;
    }

    calls = this.extractRehearsalToolCalls(cleaned);
    if (calls.length > 0) {
      return calls;
    }

    return this.extractQwenXmlToolCalls(cleaned);
  }

  private extractJsonToolCalls(text: string): ToolCall[] {
    const cleaned = text.replace(this.codeFencePattern, "").trim();

    const calls: ToolCall[] = [];
    let i = 0;
    while (i < cleaned.length) {
      if (cleaned[i] === "{") {
        let depth = 0;
        let j = i;
        while (j < cleaned.length) {
          if (cleaned[j] === "{") {
            depth++;
          } else if (cleaned[j] === "}") {
            depth--;
            if (depth === 0) {
              const candidate = cleaned.slice(i, j + 1);
              const call = this.tryParseToolCall(candidate);
              if (call) {
                calls.push(call);
              }
              i = j + 1;
              break;
            }
          }
          j++;
        }
        if (depth !== 0) {
          i++;
        }
      } else {
        i++;
      }
    }

    return calls;
  }

  private tryParseToolCall(jsonStr: string): ToolCall | null {
    let data: Record<string, unknown>;
    try {
      data = JSON.parse(jsonStr);
    } catch {
      return null;
    }

    const toolName = (data.tool as string) || (data.name as string);
    if (!toolName) {
      return null;
    }

    if (!this.toolNames.has(toolName)) {
      return null;
    }

    const args = (data.args as Record<string, unknown>) ||
      (data.arguments as Record<string, unknown>) || {};

    return { tool: toolName, args };
  }

  private extractRehearsalToolCalls(text: string): ToolCall[] {
    const calls: ToolCall[] = [];
    const regex = /(\w+)\[ARGS\](\{.*?\})/g;
    let match: RegExpExecArray | null;

    while ((match = regex.exec(text)) !== null) {
      const toolName = match[1];
      const argsStr = match[2];

      if (!this.toolNames.has(toolName)) {
        continue;
      }

      try {
        const args = JSON.parse(argsStr);
        if (typeof args === "object" && args !== null) {
          calls.push({ tool: toolName, args: args as Record<string, unknown> });
        }
      } catch {
        continue;
      }
    }

    return calls;
  }

  private extractQwenXmlToolCalls(_text: string): ToolCall[] {
    return [];
  }
}

File: go/middleware/guardrails/response_validator.go
package guardrails

import (
	"encoding/json"
	"regexp"
	"strings"
)

type ToolCall struct {
	Tool string
	Args map[string]interface{}
}

type ValidationResult struct {
	ToolCalls  []ToolCall
	Nudge      *Nudge
	NeedsRetry bool
}

type ResponseValidator struct {
	ToolNames     []string
	RescueEnabled bool
	RetryNudgeFn  func(rawResponse string, toolNames []string) *Nudge
}

func NewResponseValidator(toolNames []string, rescueEnabled bool) *ResponseValidator {
	return &ResponseValidator{
		ToolNames:     toolNames,
		RescueEnabled: rescueEnabled,
		RetryNudgeFn:  RetryNudge,
	}
}

func (v *ResponseValidator) ValidateTextResponse(response string) ValidationResult {
	if v.RescueEnabled {
		rescued := v.RescueToolCall(response)
		if len(rescued) > 0 {
			return ValidationResult{
				ToolCalls:  rescued,
				Nudge:      nil,
				NeedsRetry: false,
			}
		}
	}

	return ValidationResult{
		ToolCalls:  nil,
		Nudge:      v.RetryNudgeFn(response, v.ToolNames),
		NeedsRetry: true,
	}
}

func (v *ResponseValidator) ValidateToolCalls(toolCalls []ToolCall) ValidationResult {
	unknown := make([]string, 0)
	validCalls := make([]ToolCall, 0)

	for _, tc := range toolCalls {
		if !v.isValidTool(tc.Tool) {
			unknown = append(unknown, tc.Tool)
		} else {
			validCalls = append(validCalls, tc)
		}
	}

	if len(unknown) > 0 {
		return ValidationResult{
			ToolCalls:  nil,
			Nudge:      UnknownToolNudge(unknown[0], v.ToolNames),
			NeedsRetry: true,
		}
	}

	return ValidationResult{
		ToolCalls:  validCalls,
		Nudge:      nil,
		NeedsRetry: false,
	}
}

func (v *ResponseValidator) isValidTool(name string) bool {
	for _, tn := range v.ToolNames {
		if tn == name {
			return true
		}
	}
	return false
}

func (v *ResponseValidator) RescueToolCall(response string) []ToolCall {
	cleaned := v.stripThinkTags(response)
	cleaned = v.stripPythonTag(cleaned)
	cleaned = strings.TrimSpace(cleaned)

	if cleaned == "" {
		return nil
	}

	calls := v.extractJSONToolCalls(cleaned)
	if len(calls) > 0 {
		return calls
	}

	calls = v.extractRehearsalToolCalls(cleaned)
	if len(calls) > 0 {
		return calls
	}

	calls = v.extractQwenXMLToolCalls(cleaned)
	return calls
}

func (v *ResponseValidator) stripThinkTags(text string) string {
	thinkPattern := regexp.MustCompile(`(?i)\[THINK\].*?\[/THINK\]|<think>.*?</think>`)
	return thinkPattern.ReplaceAllString(text, "")
}

func (v *ResponseValidator) stripPythonTag(text string) string {
	pythonTagPattern := regexp.MustCompile(`(?i)<\|python_tag\|>`)
	return pythonTagPattern.ReplaceAllString(text, "")
}

func (v *ResponseValidator) extractJSONToolCalls(text string) []ToolCall {
	codeFencePattern := regexp.MustCompile("```(?:json)?\\s*\\n?")
	cleaned := codeFencePattern.ReplaceAllString(text, "")

	cleaned = strings.Trim(cleaned, " \n\t")

	found := make([]ToolCall, 0)
	i := 0
	for i < len(cleaned) {
		if cleaned[i] == '{' {
			depth := 0
			var j int
			for j = i; j < len(cleaned); j++ {
				if cleaned[j] == '{' {
					depth++
				} else if cleaned[j] == '}' {
					depth--
					if depth == 0 {
						candidate := cleaned[i : j+1]
						if call := v.tryParseToolCall(candidate); call != nil {
							found = append(found, *call)
						}
						i = j + 1
						break
					}
				}
			}
			if depth != 0 {
				i++
			}
		} else {
			i++
		}
	}
	return found
}

func (v *ResponseValidator) tryParseToolCall(jsonStr string) *ToolCall {
	var data map[string]interface{}
	if err := json.Unmarshal([]byte(jsonStr), &data); err != nil {
		return nil
	}

	toolName, ok := data["tool"].(string)
	if !ok {
		toolName, ok = data["name"].(string)
	}
	if !ok {
		return nil
	}

	if !v.isValidTool(toolName) {
		return nil
	}

	args, ok := data["args"].(map[string]interface{})
	if !ok {
		args, ok = data["arguments"].(map[string]interface{})
		if !ok {
			args = make(map[string]interface{})
		}
	}

	return &ToolCall{Tool: toolName, Args: args}
}

var rehearsalPattern = regexp.MustCompile(`(\w+)\[ARGS\](\{.*\})`)

func (v *ResponseValidator) extractRehearsalToolCalls(text string) []ToolCall {
	found := make([]ToolCall, 0)

	matches := rehearsalPattern.FindAllStringSubmatch(text, -1)
	for _, match := range matches {
		if len(match) != 3 {
			continue
		}

		toolName := match[1]
		argsStr := match[2]

		if !v.isValidTool(toolName) {
			continue
		}

		var args map[string]interface{}
		if err := json.Unmarshal([]byte(argsStr), &args); err != nil {
			continue
		}

		if args != nil {
			found = append(found, ToolCall{Tool: toolName, Args: args})
		}
	}

	return found
}

func (v *ResponseValidator) extractQwenXMLToolCalls(text string) []ToolCall {
	return []ToolCall{}
}
